### PR TITLE
fix(core): add workspaces path if package path is not included

### PR DESCRIPTION
**Pull request by xiongemi**
## Current Behavior

- when import to a path where is not in the workspaces, it currently just shows a warning. however, it will cause an error like "module not found" because there are packages not installed.

## Expected Behavior

- automatically add to the workspaces
Screenshot 2024-11-08 at 12 58 42 AM

## Related Issue(s)

Fixes #

### CODE REVIEW

File: packages/nx/src/utils/package-manager.ts [hunk 1]
```diff
@@ -478,3 +481,76 @@ export async function packageRegistryPack(
...
+    const updatedPackageJson = {
+      ...packageJson,
+      workspaces,
+    };
+    const packageJsonPath = join(root, 'package.json');
+    writeJsonFile(packageJsonPath, updatedPackageJson);
+  } else if (packageManager === 'pnpm') {
+    const yamlPath = join(root, 'pnpm-workspace.yaml');
+    let yaml: any = {};
+    if (existsSync(yamlPath)) {
```
Comment (jaysoo) [inline packages/nx/src/utils/package-manager.ts:549]: Would this preserve YAML comments? I haven't tried it yet.

File: packages/nx/src/command-line/import/import.ts [hunk 2]
```diff
@@ -391,76 +392,66 @@ async function createTemporaryRemote(
...
+  let added = false;
+  try {
+    added = addPackagePathToWorkspaces(pm, workspaceRoot, pkgPath);
+  } catch {
+    return;
+  }
+  if (!added) {
+    return;
+  }
+  await destinationGitClient.amendCommit();
```
Comment (FrozenPandaz) [inline packages/nx/src/command-line/import/import.ts:412]: I think this should all be under the `isWorkspacesEnabled` section right?

File: packages/nx/src/command-line/import/import.ts [hunk 3]
```diff
@@ -391,76 +392,66 @@ async function createTemporaryRemote(
...
           ? [
               `We recommend enabling Bun workspaces to install dependencies for the imported project.`,
-              `Add \`"workspaces": ["${pkgPath}"]\` to package.json and run "${pmc.install}".`,
+              `Added \`"workspaces": ["${pkgPath}"]\` to package.json.`,
               `See: https://bun.sh/docs/install/workspaces`,
             ]
           : [
               `We recommend enabling PNPM workspaces to install dependencies for the imported project.`,
-              `Add the following entry to to pnpm-workspace.yaml and run "${pmc.install}":`,
+              `Added the following entry to to pnpm-workspace.yaml:`,
```
Comment (FrozenPandaz) [inline packages/nx/src/command-line/import/import.ts:437]: I don't think this should change. If someone doesn't have `npm workspaces` setup, we shouldn't enable it for them. We should keep it as is and tell them to enable it altogether. If someone does have `npm workspaces` enabled, then that's when we should add it for them.

File: packages/nx/src/command-line/import/import.ts [hunk 4]
```diff
@@ -391,76 +392,66 @@ async function createTemporaryRemote(
...
+      bodyLines:
+        pm === 'npm' || pm === 'yarn' || pm === 'bun'
+          ? [
+              `The imported project (${pkgPath}) is missing the "workspaces" field in package.json.`,
+              `Added "${pkgPath}" to workspaces.`,
+            ]
+          : [
+              `The imported project (${pkgPath}) is missing the "packages" field in pnpm-workspaces.yaml.`,
+              `Added "${pkgPath}" to packages.`,
+            ],
```
Comment (FrozenPandaz) [inline packages/nx/src/command-line/import/import.ts:454]: Can we make this a success message instead of a warning? It should tell people that nx has successfully added the path to the workspaces/packages field.

**Review by FrozenPandaz — CHANGES_REQUESTED** [review]

File: packages/nx/src/command-line/import/import.ts (hunk 4 above)
Comment (xiongemi) [reply packages/nx/src/command-line/import/import.ts:454]: changed

File: packages/nx/src/command-line/import/import.ts (hunk 3 above)
Comment (xiongemi) [reply packages/nx/src/command-line/import/import.ts:437]: changed back

File: packages/nx/src/command-line/import/import.ts (hunk 2 above)
Comment (xiongemi) [reply packages/nx/src/command-line/import/import.ts]: changed

File: packages/nx/src/utils/package-manager.ts (hunk 1 above)
Comment (xiongemi) [reply packages/nx/src/utils/package-manager.ts]: changed the code, it now preserves comments

File: packages/nx/src/command-line/import/import.ts [hunk 5]
```diff
@@ -280,21 +284,44 @@ export async function importHandler(options: ImportOptions) {
...
   // If install fails, we should continue since the errors could be resolved later.
   let installFailed = false;
   if (plugins.length > 0) {
+    output.log({ title: 'Installing Plugins' });
     try {
-      output.log({ title: 'Installing Plugins' });
-      installPlugins(workspaceRoot, plugins, pmc, updatePackageScripts);
-
-      await destinationGitClient.amendCommit();
+      runPackageManagerInstallPlugins(workspaceRoot, pmc, plugins);
```
Comment (FrozenPandaz) [inline packages/nx/src/command-line/import/import.ts:299]: This is unrelated.

File: packages/nx/src/command-line/import/import.ts [hunk 6]
```diff
@@ -428,39 +455,47 @@ async function warnOnMissingWorkspacesEntry(
...
-      const packageJson = readPackageJson();
-      workspaces = packageJson.workspaces;
-    } else if (pm === 'pnpm') {
-      const yamlPath = join(workspaceRoot, 'pnpm-workspace.yaml');
-      if (existsSync(yamlPath)) {
-        const yamlContent = await fsp.readFile(yamlPath, 'utf-8');
-        const yaml = yamlLoad(yamlContent);
-        workspaces = yaml.packages;
-      }
+    let workspaces: string[] = getPackageWorkspaces(pm, workspaceRoot) ?? [];
```
Comment (FrozenPandaz) [inline packages/nx/src/command-line/import/import.ts:458]: getPackageWorkspaces should at least return an empty array if workspaces is indeed enabled.

File: packages/nx/src/command-line/import/import.ts [hunk 7]
```diff
@@ -428,39 +455,47 @@ async function warnOnMissingWorkspacesEntry(
...
-                  `Add "${pkgsDir}/*" to packages run "${pmc.install}".`,
-                ],
-        });
-      }
+    let added = false;
+    try {
+      added = addPackagePathToWorkspaces(pm, workspaceRoot, pkgPath);
+    } catch {
+      // ignore error
+    }
```
Comment (FrozenPandaz) [inline packages/nx/src/command-line/import/import.ts:469]: Why would this error?

File: packages/nx/src/command-line/init/init-v2.ts [hunk 8]
```diff
@@ -36,33 +37,73 @@ export interface InitArgs {
   integrated?: boolean; // For Angular projects only
 }
 
-export function installPlugins(
+export function runPackageManagerInstallPlugins(
```
Comment (FrozenPandaz) [inline packages/nx/src/command-line/init/init-v2.ts:40]: This is unrelated

File: packages/nx/src/utils/package-manager.ts [hunk 9]
```diff
@@ -478,3 +494,120 @@ export async function packageRegistryPack(
...
+    packageManager === 'bun'
+  ) {
+    const packageJson = readPackageJson(root);
+    workspaces = packageJson.workspaces;
+  } else if (packageManager === 'pnpm') {
+    const pnpmWorkspacePath = join(root, 'pnpm-workspace.yaml');
+    if (existsSync(pnpmWorkspacePath)) {
+      const { packages } =
+        readYamlFile<{ packages: string[] }>(pnpmWorkspacePath) ?? {};
+      workspaces = packages;
```
Comment (FrozenPandaz) [inline packages/nx/src/utils/package-manager.ts:520]:
```suggestion
      workspaces = packages ?? [];
```

File: packages/nx/src/utils/package-manager.ts [hunk 10]
```diff
@@ -478,3 +494,120 @@ export async function packageRegistryPack(
...
+export function addPackagePathToWorkspaces(
+  packageManager: PackageManager = detectPackageManager(),
+  root: string = workspaceRoot,
+  packagePath: string
+): boolean {
+  let workspaces: string[] = getPackageWorkspaces(packageManager, root) ?? [];
+  const isPkgIncluded = workspaces.some((w) => minimatch(packagePath, w));
+  if (isPkgIncluded) {
+    return false;
+  }
```
Comment (FrozenPandaz) [inline packages/nx/src/utils/package-manager.ts:544]: Can we not check this.. and always add it? It's up to the user of this function to check if it already exists.And then this function does not need to return a boolean.

File: packages/nx/src/utils/package-manager.ts [hunk 11]
```diff
@@ -478,3 +494,120 @@ export async function packageRegistryPack(
...
+      writeFileSync(
+        pnpmWorkspacePath,
+        YAMLStringify({
+          packages: [packagePath],
+        })
+      );
+    }
+    return true;
+  }
+  return false;
```
Comment (FrozenPandaz) [inline packages/nx/src/utils/package-manager.ts:612]: All the package managers are handled right?

**Review by FrozenPandaz — CHANGES_REQUESTED** [review]

File: packages/nx/src/command-line/init/init-v2.ts (hunk 8 above)
Comment (xiongemi) [reply packages/nx/src/command-line/init/init-v2.ts]: removed

File: packages/nx/src/utils/package-manager.ts (hunk 9 above)
Comment (xiongemi) [reply packages/nx/src/utils/package-manager.ts:520]: changed

**Review by FrozenPandaz — APPROVED** [review]
